File: csv-tools/scripts/stats.py
#!/usr/bin/env python3
# Copyright 2026 Yevgeniy Zaremba
# SPDX-License-Identifier: Apache-2.0

"""Per-column statistics: presence, unique count, top values, numeric min/max/mean."""

from __future__ import annotations

import argparse
from collections import Counter

from common import load_csv, parse_delimiter, sniff_type, write_json


def add_dialect_args(parser: argparse.ArgumentParser) -> None:
    parser.add_argument("input", nargs="?", default="-", help="Input CSV file or '-' for stdin.")
    parser.add_argument("--delimiter", default=",", help="Field delimiter (use '\\t' for tab).")
    parser.add_argument("--no-header", action="store_true", help="First row is data.")
    parser.add_argument("--skip-lines", type=int, default=None, help="Skip this many preamble lines; next line is header.")
    parser.add_argument("--comment-char", help="Skip lines starting with this character.")
    parser.add_argument("--encoding", default="utf-8", help="File encoding.")


def parse_fields(raw: str | None) -> list[str]:
    if not raw:
        return []
    return [f.strip() for f in raw.split(",") if f.strip()]


def numeric_summary(values: list[str]) -> dict | None:
    nums: list[float] = []
    for v in values:
        v = (v or "").strip()
        if not v:
            continue
        try:
            nums.append(float(v))
        except ValueError:
            pass
    if not nums:
        return None
    return {
        "count": len(nums),
        "min": min(nums),
        "max": max(nums),
        "mean": sum(nums) / len(nums),
    }


def main() -> None:
    parser = argparse.ArgumentParser(description="Compute per-column statistics for CSV.")
    add_dialect_args(parser)
    parser.add_argument("--fields", help="Comma-separated columns to analyze (default: all).")
    parser.add_argument("--top", type=int, default=10, help="Top N frequent values per column.")
    parser.add_argument("--compact", action="store_true", help="Emit compact JSON.")
    args = parser.parse_args()

    columns, rows = load_csv(
        args.input,
        delimiter=parse_delimiter(args.delimiter),
        has_header=not args.no_header,
        comment_char=args.comment_char,
        encoding=args.encoding,
        skip_lines=args.skip_lines,
    )[:2]

    selected = parse_fields(args.fields) if args.fields else columns
    selected = [c for c in selected if c in columns]
    if not selected:
        selected = columns

    field_stats: dict = {}
    for col in selected:
        values = [row.get(col, "") or "" for row in rows]
        presence = sum(1 for v in values if (v or "").strip())
        freq = Counter(values)
        entry = {
            "presence": f"{presence}/{len(rows)}",
            "unique_values": len(freq),
            "top_values": [{"value": k, "count": c} for k, c in freq.most_common(max(0, args.top))],
        }
        num = numeric_summary(values)
        if num:
            entry["numeric"] = num
        field_stats[col] = entry

    result = {"record_count": len(rows), "field_count": len(selected), "fields": field_stats}
    write_json(result, compact=args.compact)


if __name__ == "__main__":
    main()

File: csv-tools/scripts-node/stats.mjs
#!/usr/bin/env node
// Copyright 2026 Yevgeniy Zaremba
// SPDX-License-Identifier: Apache-2.0

import { parseArgs } from "node:util";
import {
  DEFAULT_ENCODING,
  loadCsv,
  parseDelimiter,
  writeJson,
} from "./common.mjs";

function parseFields(raw) {
  if (!raw) return [];
  return raw.split(",").map((f) => f.trim()).filter(Boolean);
}

function numericSummary(values) {
  const nums = [];
  for (const v of values) {
    const s = (v ?? "").trim();
    if (!s) continue;
    const n = Number.parseFloat(s);
    if (!Number.isNaN(n)) nums.push(n);
  }
  if (!nums.length) return null;
  const sum = nums.reduce((a, b) => a + b, 0);
  return {
    count: nums.length,
    min: Math.min(...nums),
    max: Math.max(...nums),
    mean: sum / nums.length,
  };
}

function main() {
  const { values, positionals } = parseArgs({
    options: {
      delimiter: { type: "string", default: "," },
      "no-header": { type: "boolean", default: false },
      "skip-lines": { type: "string" },
      "comment-char": { type: "string" },
      encoding: { type: "string", default: DEFAULT_ENCODING },
      fields: { type: "string" },
      top: { type: "string", default: "10" },
      compact: { type: "boolean", default: false },
    },
    allowPositionals: true,
  });

  const input = positionals[0] ?? "-";
  const { columns, rows } = loadCsv(input, {
    delimiter: parseDelimiter(values.delimiter),
    hasHeader: !values["no-header"],
    commentChar: values["comment-char"] ?? null,
    encoding: values.encoding,
    skipLines: values["skip-lines"] != null ? Number.parseInt(values["skip-lines"], 10) : null,
  });

  let selected = values.fields ? parseFields(values.fields) : columns;
  selected = selected.filter((c) => columns.includes(c));
  if (!selected.length) selected = [...columns];

  const topN = Math.max(0, Number.parseInt(values.top, 10));
  const fieldStats = {};

  for (const col of selected) {
    const colValues = rows.map((r) => r[col] ?? "");
    const presence = colValues.filter((v) => (v || "").trim()).length;
    const freq = new Map();
    for (const v of colValues) {
      freq.set(v, (freq.get(v) ?? 0) + 1);
    }
    const sorted = [...freq.entries()].sort((a, b) => b[1] - a[1]).slice(0, topN);
    const entry = {
      presence: `${presence}/${rows.length}`,
      unique_values: freq.size,
      top_values: sorted.map(([value, count]) => ({ value, count })),
    };
    const num = numericSummary(colValues);
    if (num) entry.numeric = num;
    fieldStats[col] = entry;
  }

  const result = {
    record_count: rows.length,
    field_count: selected.length,
    fields: fieldStats,
  };
  writeJson(result, values.compact);
}

main();
